Documentation/design-docs/standalone-gc-eventing.md

```diff
@@ -0,0 +1,237 @@
+we do want it to be easy to add new events. It will likely be difficult to balance auto-generated code with the need for
+subtle custom modifications to event dispatch code.
+
+Tools (specifically PerfView) will want to be enlightened about dynamic events fired by the runtime. The extent to which
+we want to make this experience nice is up to us, but we will most likely want to ship PerfView support for any dynamic event
+that we end up shipping with the CLR GC.
+
+[1]: https://github.com/dotnet/coreclr/blob/cab0db6345a7941f75d991281bcc0079d28ba182/src/gc/env/etmdummy.h#L5-L57 
+
+### Concrete Example: Porting a single known event to IGCToCLREventSink
+
+The following steps illustrate what needs to be done to bring a single event over to `IGCToCLREventSink`:
+
+Two things are needed by the GC in order to fire an event: a way to determine if the event is on, and a way to fire
+the event. Events can be fired even if they aren't enabled (they occasionally are); the platform loggers will ignore the
+event if it is not enabled. However, the GC generally avoids doing expensive eventing-related operations if an event is
+not on.
+
+The GC generally uses the `ETW_EVENT_ENABLED` macro to query whether an event is on. `ETW_EVENT_ENABLED` will be
+implemented in terms of `GCEventStatus::IsEnabled` above, so you will need to define appropriate macros in order for your
+event to work here. This will likely mean that you will need to define a macro that turns the name of your event into a
+pair of a level and a keyword, which will determine whether or not your event is enabled.
+
+To fire the event, you can add your event's callback to `IGCToCLREventSink`:
+
+```c++
+class IGCToCLREventSink
+{
+    virtual void FireYourEvent(
+        /* your event arguments here... */
+    ) = 0;
+};
+```
+
+You can then define the `FireYourEvent` macro in `src/gc/env/etmdummy.h` to point to your new method on `IGCToCLREventSink`
+Your implementation of `FireYourEvent` in the EE will need to calculate any EE-specific data (e.g. `ClrInstanceId`) and
+then forward the event arguments onto the platform logger, which can be done with the `Fire` macros that the EE has access
+to (that we implemented for the GC).
+
+### Concrete Example: Adding a dynamic event
+
+The following steps illustrate what needs to be done to add a new dynamic event:
+
+There are two things that need to be written: the `ETW_EVENT_ENABLED` support for your new event and the macro responsible
+for firing your event. `ETW_EVENT_ENABLED` can be implemented in the same manner as a known event, by introducing a macro
+for your event name that expands to your event's level and keyword.
+
+Firing of a dynamic event ultimately must call `FireDynamicEvent` with an event name and a serialized payload. Therefore,
+it is the responsibility of the GC to format an event's payload into a binary format. It is ultimately up to you(1) to write
```

Comment:
but we are going with CTF for the format?

Reply:
no - the binary format is 100% controlled by us and can even be e.g. json if we wanted to. CTF is way below us in terms of abstraction at this level, I think.

Reply:
Maoni and I followed-up on this offline